packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -570,21 +603,37 @@ export class ConnectionManager implements IConnectionManager {
         return true;
     }
 
+    /**
+     * Cancel in-progress connection attempt.
+     */
+    private cancelConnection() {
+        assert(this.pendingConnection !== undefined, "this.pendingConnection is undefined when trying to cancel");
+        this.pendingConnection.abortController.abort();
+        this.pendingConnection = undefined;
+        this.logger.sendTelemetryEvent({ eventName: "ConnectionAttemptCancelled" });
+    }
+
     /**
      * Once we've successfully gotten a connection, we need to set up state, attach event listeners, and process
      * initial messages.
      * @param connection - The newly established connection
      */
-     private setupNewSuccessfulConnection(connection: IDocumentDeltaConnection, requestedMode: ConnectionMode) {
+     private setupNewSuccessfulConnection(
+         connection: IDocumentDeltaConnection,
+         requestedMode: ConnectionMode,
+         abortSignal: AbortSignal,
+        ) {
         // Old connection should have been cleaned up before establishing a new one
         assert(this.connection === undefined, 0x0e6 /* "old connection exists on new connection setup" */);
         assert(!connection.disposed, 0x28a /* "can't be disposed - Callers need to ensure that!" */);
 
-        if (this.pendingConnection) {
-            this.pendingConnection = false;
-        } else {
-            assert(this.closed, 0x27f /* "reentrancy may result in incorrect behavior" */);
```

Comment:
I think that this assert is a bit outdated. With our new cancellation logic it is possible for us to reach `setupNewSuccessfulConnection()` after the connection is cancelled. In this case, `pendingConnection` would be undefined and the container would not be closed, triggering the assert.

Reply:
Ok. Can you make sure there's a test covering this case that succeeds without the assert?

Reply:
It is a bit tricky to directly test, since we need to trigger a cancellation after we exit the `while` loop, but before we complete `setupNewSuccessfulConnection()`. I did however come across this scenario while debugging other tests and it is working as expected.